app/models/mixins/miq_provision_quota_mixin.rb

```diff
@@ -295,33 +352,59 @@ def quota_provision_stats(prov_method, options)
-        num_vms_for_request = no_vm.length
-      end
+      pr.type == "ServiceTemplateProvisionRequest" ? service_quota_values(pr, result) : vm_quota_values(pr, result)
+    end
+    result
+  end
 
-      unless num_vms_for_request.zero?
-        new_disk_storage_size = pr.get_new_disks.inject(0) { |s, d| s += d[:sizeInMB].to_i } * 1.megabyte
-        result[:count] += num_vms_for_request
-        result[:memory] += pr.get_option(:vm_memory).to_i * num_vms_for_request
-        result[:cpu] += pr.get_option(:number_of_cpus).to_i * num_vms_for_request
-        result[:storage] += (pr.vm_template.allocated_disk_storage.to_i + new_disk_storage_size) * num_vms_for_request
-        result[:ids] << pr.id
-
-        # Include a resource breakdown for actively provisioning records
-        pr.miq_request_tasks.each do |p|
-          next unless p.state == 'active'
-          host_id, storage_id = p.get_option(:dest_host).to_i, p.get_option(:dest_storage).to_i
-          active = result[:active]
-          active[:memory_by_host_id][host_id] += p.get_option(:vm_memory).to_i
-          active[:cpu_by_host_id][host_id] += p.get_option(:number_of_cpus).to_i
-          active[:storage_by_id][storage_id] += p.vm_template.allocated_disk_storage.to_i + new_disk_storage_size
-          active[:vms_by_storage_id][storage_id] << p.id
-          active[:ids] << p.id
-        end
+  def number_of_vms(request)
+    num_vms_for_request = request.get_option(:number_of_vms).to_i
+    if options[:nil_vm_id_only] == true && request.miq_request_tasks.length == num_vms_for_request
+      num_vms_for_request = request.miq_request_tasks.where(:destination_id => nil).where.not(:state => 'finished').count
+    end
+    num_vms_for_request
+  end
+
+  def cloud?(request)
+    request.source.try(:cloud) || false
+  end
+
+  def vendor(request)
+    request.source.try(:vendor)
+  end
+
+  def flavor(request)
+    Flavor.find_by(:id => request.get_option(:instance_type)) if cloud?(request)
+  end
+
+  def number_of_cpus(prov, cloud, flavor_obj)
+    if cloud
+      flavor_obj.try(:cpus)
+    else
+      request = prov.kind_of?(MiqRequest) ? prov : prov.miq_request
+      num_cpus = request.get_option(:number_of_sockets).to_i * request.get_option(:cores_per_socket).to_i
+      num_cpus.zero? ? request.get_option(:number_of_cpus).to_i : num_cpus
+    end
+  end
+
+  def storage(prov, cloud, vendor, flavor_obj = nil)
+    if cloud
+      if vendor == 'google'
+        return prov.get_option(:boot_disk_size).to_i.gigabytes
       end
+      return nil unless flavor_obj
+      flavor_obj.root_disk_size.to_i + flavor_obj.ephemeral_disk_size.to_i + flavor_obj.swap_disk_size.to_i
+    else
+      prov.kind_of?(MiqRequest) ? prov.vm_template.provisioned_storage : prov.miq_request.vm_template.provisioned_storage
+    end
+  end
+
+  def memory(prov, cloud, vendor, flavor_obj = nil)
+    if cloud
```

Comment:
Use guard clauses to prevent doing extra work.
```ruby
return flavor_obj.try(:memory) if cloud
return if %w(amazon openstack google).includes?(vendor)
request = prov.kind_of?(MiqRequest) ? prov : prov.miq_request
request.get_option(:vm_memory).to_i.megabytes
```